File: Benchmarking/src/gen_8_14.cpp
  #include <vector>
  #include <iostream>
  #include <chrono>
  struct s0 {
    std::vector<double> p0;
    std::vector<std::vector<double>> p1;
    s0(std::vector<double> v0, std::vector<std::vector<double>> v1): p0(v0), p1(v1) { }
  };
  struct s1 {
    s0 p0;
    std::vector<s0> p1;
    s1(s0 v0, std::vector<s0> v1): p0(v0), p1(v1) { }
  };
  struct s6 {
    std::vector<s1> p0;
    std::vector<std::vector<s0>> p1;
    s6(std::vector<s1> v0, std::vector<std::vector<s0>> v1): p0(v0), p1(v1) { }
  };
  __attribute__((noinline))
  double f0(const s6 &v0, const std::vector<std::vector<s1>> &v1, const double &v2) {
    const std::vector<std::vector<s0>> v7 = v0.p1;
    const std::vector<s0> v20 = v7[1];
    const double v14 = v2 + v2;
    const s0 v18 = v20[0];
    double v32 = v2;
    const double v23 = v2 * v2;
    const double v47 = v23 + v32;
    const std::vector<std::vector<double>> v34 = v18.p1;
    const std::vector<double> v46 = v34[2];
    v32 = v47;
    std::vector<double> v36 = v46;
    v36[0] = v14;
    v36[0] = v32;
    const double v72 = v36[0];
    double v191 = v72;
    return v191;
  }
  int main() {
    s6 v0({ { { { { 0.0, 1.0, 2.0 }, { { 3.0 }, { 4.0 }, { 5.0 } } }, { { { 6.0, 7.0, 8.0 }, { { 9.0 }, { 10.0 }, { 11.0 } } } } }, { { { 12.0, 13.0, 14.0 }, { { 15.0 }, { 16.0 }, { 17.0 } } }, { { { 18.0, 19.0, 20.0 }, { { 21.0 }, { 22.0 }, { 23.0 } } } } }, { { { 24.0, 25.0, 26.0 }, { { 27.0 }, { 28.0 }, { 29.0 } } }, { { { 30.0, 31.0, 32.0 }, { { 33.0 }, { 34.0 }, { 35.0 } } } } } }, { { { { 36.0, 37.0, 38.0 }, { { 39.0 }, { 40.0 }, { 41.0 } } } }, { { { 42.0, 43.0, 44.0 }, { { 45.0 }, { 46.0 }, { 47.0 } } } }, { { { 48.0, 49.0, 50.0 }, { { 51.0 }, { 52.0 }, { 53.0 } } } } } });
    std::vector<std::vector<s1>> v1({ { { { { 54.0, 55.0, 56.0 }, { { 57.0 }, { 58.0 }, { 59.0 } } }, { { { 60.0, 61.0, 62.0 }, { { 63.0 }, { 64.0 }, { 65.0 } } } } } } });
    double v2(66.0);
    auto start = std::chrono::high_resolution_clock::now();
    double result;
    for (int i = 0; i < 1000; i ++) {
      result = f0(v0, v1, v2);
    }
    auto end = std::chrono::high_resolution_clock::now();
    double value = *((double*) &result);
    std::cout << value << "\n";
    std::cout << std::chrono::duration_cast<std::chrono::nanoseconds>(end-start).count();
    std::cout << "\n";
    return 0;
  }

File: Benchmarking/src/gen_8_14.scala
import java.lang.System.nanoTime
import scala.collection.immutable.Vector
object Gen extends App {
  case class s0 (
    p0: Vector[Double],
    p1: Vector[Vector[Double]]
  )
  case class s1 (
    p0: s0,
    p1: Vector[s0]
  )
  case class s6 (
    p0: Vector[s1],
    p1: Vector[Vector[s0]]
  )
  @noinline
  def f0(v0: s6, v1: Vector[Vector[s1]], v2: Double): Double = {
    val v7: Vector[Vector[s0]] = v0.p1
    val v20: Vector[s0] = v7(1)
    val v14: Double = v2 + v2
    val v18: s0 = v20(0)
    var v32: Double = v2
    val v23: Double = v2 * v2
    val v47: Double = v23 + v32
    val v34: Vector[Vector[Double]] = v18.p1
    val v46: Vector[Double] = v34(2)
    v32 = v47
    var v36: Vector[Double] = v46
    v36 = v36.updated(0, v14)
    v36 = v36.updated(0, v32)
    val v72: Double = v36(0)
    var v191: Double = v72
    v191
  }
  def benchmark(): Unit = {
    val v0: s6 = s6(Vector(s1(s0(Vector(0.0, 1.0, 2.0), Vector(Vector(3.0), Vector(4.0), Vector(5.0))), Vector(s0(Vector(6.0, 7.0, 8.0), Vector(Vector(9.0), Vector(10.0), Vector(11.0))))), s1(s0(Vector(12.0, 13.0, 14.0), Vector(Vector(15.0), Vector(16.0), Vector(17.0))), Vector(s0(Vector(18.0, 19.0, 20.0), Vector(Vector(21.0), Vector(22.0), Vector(23.0))))), s1(s0(Vector(24.0, 25.0, 26.0), Vector(Vector(27.0), Vector(28.0), Vector(29.0))), Vector(s0(Vector(30.0, 31.0, 32.0), Vector(Vector(33.0), Vector(34.0), Vector(35.0)))))), Vector(Vector(s0(Vector(36.0, 37.0, 38.0), Vector(Vector(39.0), Vector(40.0), Vector(41.0)))), Vector(s0(Vector(42.0, 43.0, 44.0), Vector(Vector(45.0), Vector(46.0), Vector(47.0)))), Vector(s0(Vector(48.0, 49.0, 50.0), Vector(Vector(51.0), Vector(52.0), Vector(53.0))))))
    val v1: Vector[Vector[s1]] = Vector(Vector(s1(s0(Vector(54.0, 55.0, 56.0), Vector(Vector(57.0), Vector(58.0), Vector(59.0))), Vector(s0(Vector(60.0, 61.0, 62.0), Vector(Vector(63.0), Vector(64.0), Vector(65.0)))))))
    val v2: Double = 66.0
    val start = nanoTime()
    var result: Double = 66.0
    (1 to 1000).foreach { _ =>
      result = f0(v0, v1, v2)
    }
    val end = nanoTime()
    println(result)
    println(end - start)
  }
  benchmark()
}